File: app/src/main/java/com/bignerdranch/android/mediafiles/ui/dashboard/DashboardFragment.java
package com.bignerdranch.android.mediafiles.ui.dashboard;

import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bignerdranch.android.mediafiles.R;
import com.bignerdranch.android.mediafiles.ui.recycleView.MediaFileImageAdapter;

public class DashboardFragment extends Fragment {

    private DashboardViewModel dashboardViewModel;
    private MediaFileImageAdapter adapter;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        dashboardViewModel =
                ViewModelProviders.of(this).get(DashboardViewModel.class);
        View root = inflater.inflate(R.layout.fragment_dashboard, container, false);

        final GridLayoutManager layoutManager = new GridLayoutManager(
                getContext(), getSpanCountForLayoutMan());
        layoutManager.setOrientation(RecyclerView.VERTICAL);
        final RecyclerView localItemRecyclerView = root.findViewById(R.id.mediaFilesImageRecyclerView);
        localItemRecyclerView.setLayoutManager(layoutManager);
        adapter = new MediaFileImageAdapter();
        localItemRecyclerView.setAdapter(adapter);

        final TextView textView = root.findViewById(R.id.text_dashboard);
        dashboardViewModel.getText().observe(getViewLifecycleOwner(), new Observer<String>() {
            @Override
            public void onChanged(@Nullable String s) {
                textView.setText(s);
            }
        });

        dashboardViewModel.getLiveMediaFiles().observe(getViewLifecycleOwner(), adapter::submitList);
        return root;
    }

    private int getSpanCountForLayoutMan() {
        DisplayMetrics metric = new DisplayMetrics();
        getActivity().getWindowManager().getDefaultDisplay().getMetrics(metric);
        int width = metric.widthPixels;
        int height = metric.heightPixels;
        float density = metric.density;
        int densityDpi = metric.densityDpi;

        float imageWidth = getResources().getDimension(R.dimen.item_image_width);

        int result = (int)(width/imageWidth);
        Log.v("ShiyihuiHLNSKQ","value: " + width + " " + height + " " + density + " " + densityDpi + " " + result);
        return result;
    }
}

File: app/src/main/java/com/bignerdranch/android/mediafiles/ui/dashboard/DashboardFragment.kt
package com.bignerdranch.android.mediafiles.ui.dashboard

import android.util.DisplayMetrics;
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.paging.PagedList
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bignerdranch.android.mediafiles.DTAG
import com.bignerdranch.android.mediafiles.MediaFilesApplication
import com.bignerdranch.android.mediafiles.R
import com.bignerdranch.android.mediafiles.discovery.model.MediaFile
import com.bignerdranch.android.mediafiles.ui.recycleView.MediaFileImageAdapter
import com.bignerdranch.android.mediafiles.util.log.Logger
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class DashboardFragment : Fragment() {
    lateinit var dashboardViewModel: DashboardViewModel
    lateinit var adapter: MediaFileImageAdapter
    lateinit var logger: Logger

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        logger = MediaFilesApplication.appComponent.getLogger()
        dashboardViewModel = ViewModelProvider(this).get(DashboardViewModel::class.java)
        val root = inflater.inflate(R.layout.fragment_dashboard, container, false)
//        val layoutManager = GridLayoutManager(context, 3)
        val layoutManager = GridLayoutManager(context, spanCountForLayoutMan)
        layoutManager.orientation = RecyclerView.VERTICAL
        val localItemRecyclerView: RecyclerView = root.findViewById(R.id.mediaFilesImageRecyclerView)
        localItemRecyclerView.layoutManager = layoutManager
        adapter = MediaFileImageAdapter(logger)
        localItemRecyclerView.adapter = adapter
        val textView = root.findViewById<TextView>(R.id.text_dashboard)
        dashboardViewModel.text.observe(viewLifecycleOwner, { s -> textView.text = s })

        lifecycleScope.launch {
            // We repeat on the STARTED lifecycle because an Activity may be PAUSED
            // but still visible on the screen, for example in a multi window app
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                dashboardViewModel.liveMediaFiles.collectLatest {
                    adapter.submitData(it)
                }
            }
        }

        return root
    }

    private val spanCountForLayoutMan: Int
        get() {
            val metric = DisplayMetrics()
            var result = 3

            activity?.let {
                it.windowManager.defaultDisplay.getMetrics(metric)
                val width = metric.widthPixels
                val height = metric.heightPixels
                val density = metric.density
                val densityDpi = metric.densityDpi
                val imageWidth = resources.getDimension(R.dimen.item_image_width)
                result = (width / imageWidth).toInt()
                logger.v(DTAG, "value: $width $height $density $densityDpi $result")
            }

            return result
        }
}